include/wil/com_apartment_variable.h

```diff
@@ -0,0 +1,230 @@
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+#ifndef __WIL_COM_APARTMENT_VARIABLE_INCLUDED
+#define __WIL_COM_APARTMENT_VARIABLE_INCLUDED
+
+#include <unordered_map>
+#include <any>
+#include <type_traits>
+#include <Unknwn.h>
+#include <winrt/base.h>
+#include <roapi.h>
+#include <objidl.h>
+#include <wil/result_macros.h>
+
+#ifndef WIL_ENABLE_EXCEPTIONS
+#error This header requires exceptions
+#endif
+
+namespace wil
+{
+    // Determine if apartment variables are supported in the current process context.
+    // Prior to https://microsoft.visualstudio.com/OS/_git/os.2020/pullrequest/5750493 the
+    // APIs needed to implement apartment variables did not work in non-packaged processes.
+    inline bool are_apartment_variables_supported()
+    {
+        unsigned long long apartmentId{};
+        return RoGetApartmentIdentifier(&apartmentId) != HRESULT_FROM_WIN32(ERROR_API_UNAVAILABLE);
+    }
+
+    namespace details
+    {
+        inline static winrt::slim_mutex s_lock;
+        // Apartment id -> variable storage. Variables are stored using the address of
+        // the function that produces the value as the key.
+        inline static std::unordered_map<unsigned long long, std::unordered_map<const void*, std::any>> s_apartmentStorage;
+    }
+
+    // Get the apartment variable, returning the value stored in the per apartment
+    // storage or use the create function to create and store it in the current apartment.
+    // This variables lifetime is tied to the COM apartment lifetime and will
+    // be safely cleaned up. Call reset_for_current_com_apartment if you want to
+    // rundown the variable before the apartment exits.
+    template<typename F>
+    auto get_for_current_com_apartment(const F& createFn) -> std::invoke_result_t<F>
+    {
+        static_assert(std::is_function_v<F>, "createFn must be a function");
+
+        unsigned long long apartmentId{};
+        FAIL_FAST_IF_FAILED(RoGetApartmentIdentifier(&apartmentId));
+
+        const auto variableKey = std::addressof(createFn);
+
+        auto lock = winrt::slim_lock_guard(details::s_lock);
+
+        auto storage = details::s_apartmentStorage.find(apartmentId);
+        if (storage != details::s_apartmentStorage.end())
+        {
+            auto& variables = storage->second;
+            auto variable = variables.find(variableKey);
+            if (variable != variables.end())
+            {
+                return std::any_cast<std::invoke_result_t<F>>(variable->second);
+            }
+            else
+            {
+                auto result = createFn();

```

Comment:
Calling creation function under the lock is asking for trouble if the creation function itself calls `get_for_current_com_apartment`!

```cpp
auto GetDefaultWidgetSite() { return WidgetSite(); }

auto GetDefaultWidget() {
    return make_shared<Widget>(wil::get_for_current_com_apartment(GetDefaultWidgetSite));
}
```

Reply:
true, but the deadlock will be deterministic and easy to diagnose. I'm going to choose the simpler design as it stands unless you feel strongly.